include/material/norsand.tcc

```diff
@@ -0,0 +1,693 @@
+
+  // Compute and update psi
+  double psi_image = void_ratio - e_image;
+  (*state_vars).at("psi_image") = psi_image;
+
+  return true;
+}
+
+//! Compute elastic tensor
+template <unsigned Tdim>
+bool mpm::NorSand<Tdim>::compute_p_bond(mpm::dense_map* state_vars) {
+
+  // Compute current zeta cohesion
+  double zeta_cohesion = exp(-m_cohesion_ * (*state_vars).at("epds"));
+
+  if (zeta_cohesion > 1.)
+    (*state_vars).at("zeta_cohesion") = 1.;
+  else if (zeta_cohesion < 1.0E-15)
+    (*state_vars).at("zeta_cohesion") = 0.;
+  else
+    (*state_vars).at("zeta_cohesion") = zeta_cohesion;
+
+  // Update p_cohesion
+  double p_cohesion = p_cohesion_initial_ * (*state_vars).at("zeta_cohesion");
+  (*state_vars).at("p_cohesion") = p_cohesion;
+
+  // Compute current zeta dilation
+  double zeta_dilation = exp(-m_dilation_ * (*state_vars).at("epds"));
+
+  if (zeta_dilation > 1.)
+    (*state_vars).at("zeta_dilation") = 1.;
+  else if (zeta_dilation < 1.0E-15)
+    (*state_vars).at("zeta_dilation") = 0.;
+  else
+    (*state_vars).at("zeta_dilation") = zeta_dilation;
+
+  // Update p_dilation
+  double p_dilation = p_dilation_initial_ * (*state_vars).at("zeta_dilation");
+  (*state_vars).at("p_dilation") = p_dilation;
+
+  return true;
+}
+
+//! Compute yield function and yield state
+template <unsigned Tdim>
+typename mpm::NorSand<Tdim>::FailureState
+    mpm::NorSand<Tdim>::compute_yield_state(double* yield_function,
+                                            const Vector6d& stress,
+                                            const mpm::dense_map* state_vars) {
+
+  // Get stress invariants
+  // Compute mean stress p
+  double mean_p = (stress(0) + stress(1) + stress(2)) / 3.;
+  if (mean_p < 1.0E-15) mean_p = 1.0E-15;
+
+  // Compute J2
+  double j2 = (std::pow((stress(0) - stress(1)), 2) +
+               std::pow((stress(1) - stress(2)), 2) +
+               std::pow((stress(0) - stress(2)), 2)) /
+                  6.0 +
+              std::pow(stress(3), 2);
+  if (Tdim == 3) j2 += std::pow(stress(4), 2) + std::pow(stress(5), 2);
+
+  // Compute q
+  double deviatoric_q = std::sqrt(3 * j2);
+  if (deviatoric_q < 1.0E-15) deviatoric_q = 1.0E-15;
+
+  // Get image pressure and M_theta
+  const double p_image = (*state_vars).at("p_image");
+  const double M_theta = (*state_vars).at("M_theta");
+  const double p_cohesion = (*state_vars).at("p_cohesion");
+  const double p_dilation = (*state_vars).at("p_dilation");
+
+  // Initialise yield status (0: elastic, 1: yield)
+  auto yield_type = FailureState::Elastic;
+
+  // Compute yield functions
+  if (bond_model_) {
+    (*yield_function) =
+        deviatoric_q / (mean_p + p_cohesion) -
+        M_theta / N_ *
+            (1 + (N_ - 1) * std::pow(((mean_p + p_cohesion) /
+                                      (p_image + p_cohesion + p_dilation)),
+                                     (N_ / (1 - N_))));
+  } else {
+    (*yield_function) =
+        deviatoric_q / mean_p -
```

Comment:
if we have a `non-bond` model could we just `p_cohesion` and `p_dilation` to zero, we then don't have to go through so many if conditions

Reply:
Alright. I can do it!

Reply:
Done so, except for a few derivatives needing if statements